jumpscale/sals/nettools/__init__.py

```diff
@@ -32,117 +47,167 @@ def tcp_connection_test(ipaddr: str, port: int, timeout: Optional[int]):
-    finally:
-        if conn:
-            conn.close()
-    return True
+        conn = socket.create_connection((ipaddr, port), timeout)
+    except (socket.gaierror, socket.herror):
+        # raised for address-related errors
+        raise
+    except OSError:
+        # (ConnectionRefusedError, socket.timeout, OSError)
+        return False
+    else:
+        conn.close()
+        return True
 
 
-def udp_connection_test(ipaddr: str, port: int, timeout=1, message=b"PING"):
-    """tests udp connection on specified port by sending specified message
+def udp_connection_test(ipaddr: str, port: int, timeout: Optional[int] = 1, message: Optional[bytes] = b"") -> bool:
+    """tests udp connection on specified port by sending specified message and expecting
+    to receive at least one byte from the socket as an indicator of connection success
 
     Args:
         ipaddr (str): ip address
         port (int): port number
         timeout (int, optional): time before the connection test fails. Defaults to None.
         message (str, optional): message to send. Defaults to b"PING"
 
+    Raises:
+        ValueError: raises if invalid ip address was used
+
     Returns:
         bool: True if the test succeeds, False otherwise
     """
-    conn = socket.socket(type=socket.SOCK_DGRAM)
+    try:
+        ip = ipaddress.ip_address(ipaddr)
+    except ValueError:
+        raise
+    if ip.version == 4:
+        sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+    else:  # create IPv6 socket when we connect to IPv6 address
+        sock = socket.socket(socket.AF_INET6, socket.SOCK_DGRAM)
     if timeout:
-        conn.settimeout(timeout)
+        sock.settimeout(timeout)
     try:
-        conn.connect((ipaddr, port))
-    except BaseException:
-        conn.close()
+        sock.sendto(message, (ipaddr, port))
+        # expecting to receive at least one byte from the socket as indication to succeed connection
+        data, _ = sock.recvfrom(1)
+    except socket.timeout:
         return False
-
-    conn.send(message)
-
-    try:
-        conn.recvfrom(8192)
-    except Exception:
+    except OSError:
         return False
+    finally:
+        sock.close()
     return True
 
 
-def wait_connection_test(ipaddr: str, port: int, timeout=5):
+def wait_connection_test(ipaddr: str, port: int, timeout: Optional[int] = 5) -> bool:
     """Will wait until port listens on the specified address
 
     Args:
-        ipaddr (str): ip address
+        ipaddr (str): ip address, or hostname
         port (int): port number
         timeout_total (int, optional): how long to wait for the connection. Defaults to 5.
 
     Returns:
         bool: True if the test succeeds, False otherwise
     """
-    port = int(port)
-    end = now().timestamp + timeout
-    while True:
-        if now().timestamp > end:
-            return False
-        if tcp_connection_test(ipaddr, port, timeout=2):
+    # port = int(port)
+    interval = int(timeout <= 2) or 2
```

Comment:
interval is integer, 2 seconds or 5 .. etc, using or will result in a boolen value (True, False) instead

Reply:
the function will use a 1-second interval if the timeout is 2 or less, otherwise, it would use 2 seconds interval.

Reply:
yeah, but not readable, it can be something like

```suggestion
    interval = 1 if timeout <= 2 else 2
```

also if i added timeout 5 seconds and each interval is 2, the final result is I'll wait 6 seconds not 5, yes ?

Reply:
yes, you are right.
do you suggest a fixed value, like 1 sec? I wanted the code to consume lower resources when waiting for a longer time. but still, fallback to one sec when the actual wait time is 2 sec, to be able to check at least twice, instead of once. so it is simple and works.
I can come up with some more complex alternatives, but still will not work in a real-time accuracy and all will have edge cases that could make the code wait a little longer.

Reply:
No problem